### PR TITLE
IVS-71/Fix GEM051&GEM52: Correct identifiers and types

### PR DESCRIPTION
**Pull request by Ghesselink**
Referring to the relating [PR](https://github.com/IfcOpenShell/IfcOpenShell/issues/5098) and [issue](https://github.com/IfcOpenShell/IfcOpenShell/issues/5098) 

## Key Issues Addressed

- **Context Type vs. Identifier Confusion**: Previously, the file mentioned in the issue was generating a list of items that represented types rather than identifiers. This mix-up affected the checks specified by GEM051 and GEM052, which should not validate identifiers or types  against these values.
 For `ContextIdentifiers`, we validate against a list of 11 specified values (or slightly more for IFC4.3). For `ContextTypes`, we validate against 'Model' and 'Plan'.

- **Correct Identifiers for IFC Versions**: The[ list referenced in the PR](https://github.com/buildingSMART/ifc-gherkin-rules/pull/249/files#diff-dbff5d219ff677cd401a6a10e53ab870b3f3a16342b5409bdd3218a329d695d1) correctly identifies valid identifiers for IFC4. [A small update has been made for IFC4X3](https://ifc43-docs.standards.buildingsmart.org/IFC/RELEASE/IFC4x3/HTML/lexical/IfcShapeRepresentation.htm)

### GEM051
- Ensure at least one context exists with a context type of either "Model" or "Plan." No other values should be allowed.
- If a subtype of IfcGeometricRepresentation lacks a ContextType, the rule passes, as the attribute is optional per [documentation](https://ifc43-docs.standards.buildingsmart.org/IFC/RELEASE/IFC4x3/HTML/lexical/IfcRepresentation.htm).

### GEM052
- Verify that each context includes at least one subcontext.
- If the subcontext has a ContextIdentifier (optional per documentation), the values must be valid. Please let me know if you think differently @Moult 
- The list of valid identifiers differs slightly between IFC4 and IFC4.3 to align closely with the documentation.

## Additional Changes
- The step implementation not also validates multiple values in the Then step (e.g., 'The value must be X or Y or Z'). Previously, only single value checks were possible.
- Test files have been updated to reflect and validate the revised scenarios

### CODE REVIEW

File: features/GEM051_Presence-of-geometric-context.feature [hunk 1]
```diff
@@ -1,26 +1,38 @@
...
-    Then Assert existence 
-    Then Its entity type is 'IfcGeometricRepresentationContext' including subtypes
+      Then Assert existence
+      Then Its entity type is 'IfcGeometricRepresentationContext' including subtypes
 
+      Examples:
+        | Schema  | Entity     |
+        | IFC2X3  | IfcProject |
+        | IFC4    | IfcContext |
+        | IFC4X3  | IfcContext |
```
Comment (evandroAlfieri) [inline features/GEM051_Presence-of-geometric-context.feature:22]:
```suggestion
        | IFC4.3  | IfcContext |
```

File: features/GEM051_Presence-of-geometric-context.feature [hunk 2]
```diff
@@ -1,26 +1,38 @@
...
+      Given Its attribute RepresentationContexts
+      Given its attribute ContextType
+
+      Then The value must be "Model" or "Plan" or "NotDefined"
+
+      Examples:
+        | Schema  | Entity     |
+        | IFC2X3  | IfcProject |
+        | IFC4    | IfcContext |
+        | IFC4X3  | IfcContext |
```
Comment (evandroAlfieri) [inline features/GEM051_Presence-of-geometric-context.feature:38]:
```suggestion
        | IFC4.3  | IfcContext |
```

File: features/GEM052_Correct-geometric-subcontexts.feature [hunk 3]
```diff
@@ -1,30 +1,29 @@
...
-Scenario: Constraints on context type - IFC4
-    Given a model with Schema "IFC4"
-    Given An IfcGeometricRepresentationSubContext
-    Given Its attribute ContextType
+        Then The values must be in '<source>'
 
-    Then The values must be in 'valid_RepresentationType_IFC4.csv'
+        Examples: 
+            | schema | source |
+            | IFC4X3 | valid_RepresentationIdentifier_IFC4.3.csv |
```
Comment (evandroAlfieri) [inline features/GEM052_Correct-geometric-subcontexts.feature:27]:
```suggestion
            | IFC4.3 | valid_RepresentationIdentifier_IFC4.3.csv |
```

File: features/GEM051_Presence-of-geometric-context.feature [hunk 4]
```diff
@@ -1,26 +1,38 @@
 @industry-practice
 @GEM
-@version1
+@version2
 @E00020
 Feature: GEM051 - Presence of Geometric Context
 The rule verifies that a geometric context is present in the model.
```
Comment (evandroAlfieri) [inline features/GEM051_Presence-of-geometric-context.feature:6]:
```suggestion
The rule verifies that a geometric context is present in the model, and has a valid ContextType.
```

File: features/GEM052_Correct-geometric-subcontexts.feature [hunk 5]
```diff
@@ -1,30 +1,29 @@
 @industry-practice
 @GEM
-@version1
+@version2
 @E00050
 Feature: GEM052 - Correct geometric subcontexts
 The rule verifies that there is a minimum of at least one subcontext per context. 
 Reference: https://github.com/buildingSMART/Sample-Test-Files/issues/137.
 The context identifier should be one of the agreed values.
```
Comment (evandroAlfieri) [inline features/GEM052_Correct-geometric-subcontexts.feature:8]:
```suggestion
The context identifier shall be one of the values in the list of allowed shape representation identifiers.
```

**Review by evandroAlfieri — CHANGES_REQUESTED** [review]

**Comment by Moult** [comment]
> The list of valid identifiers differs slightly between IFC4 and IFC4.3 to align closely with the documentation.

Yeah. I also missed Body-Fallback :)

> If the subcontext has a ContextIdentifier (optional per documentation), the values must be valid.

I'd say that the subcontext _must_ have a ContextIdentifier (even though the schema cardinality is optional) because if it is missing, the geometry is pretty meaningless.

**Comment by Ghesselink** [comment]
> because if it is missing, the geometry is pretty meaningless.
Fair point. Considering it triggers a warning instead of an error, we can derive a little from the documentation.
Both GEM051 and GEM052 are modified and yield a warning in case the attribute for checking types or identifiers are empty. Also, a couple of a testfiles are added to reflect these changes.
https://github.com/buildingSMART/ifc-gherkin-rules/pull/251/commits/723b73ff9a5f003e1f539bc22d0183c98d7aca8c

File: features/GEM051_Presence-of-geometric-context.feature [hunk 6]
```diff
@@ -1,26 +1,37 @@
 @industry-practice
 @GEM
-@version1
+@version2
 @E00020
 Feature: GEM051 - Presence of Geometric Context
-The rule verifies that a geometric context is present in the model.
+The rule verifies that a geometric context is present in the model, and has a valid ContextType.
```
Comment (evandroAlfieri) [inline features/GEM051_Presence-of-geometric-context.feature:6]:
```suggestion
The rule verifies that a geometric context is present in the model, that its attribute ContextType is provided (not empty) and its value is valid.
```

File: features/GEM052_Correct-geometric-subcontexts.feature [hunk 7]
```diff
@@ -1,30 +1,37 @@
 @industry-practice
 @GEM
-@version1
+@version2
 @E00050
 Feature: GEM052 - Correct geometric subcontexts
 The rule verifies that there is a minimum of at least one subcontext per context.
```
Comment (evandroAlfieri) [inline features/GEM052_Correct-geometric-subcontexts.feature:6]:
```suggestion
The rule verifies that there is a minimum of at least one subcontext per context, that its attribute ContextIdentifier is provided (not empty) and its value is one of the allowed values in the list of shape representation identifiers.
```

File: features/GEM052_Correct-geometric-subcontexts.feature [hunk 8]
```diff
@@ -1,30 +1,37 @@
 @industry-practice
 @GEM
-@version1
+@version2
 @E00050
 Feature: GEM052 - Correct geometric subcontexts
 The rule verifies that there is a minimum of at least one subcontext per context. 
 Reference: https://github.com/buildingSMART/Sample-Test-Files/issues/137.
-The context identifier should be one of the agreed values.
+The context identifier shall be one of the values in the list of allowed shape representation identifiers.
```
Comment (evandroAlfieri) [inline features/GEM052_Correct-geometric-subcontexts.feature:8]:
```suggestion
```

**Review by evandroAlfieri — CHANGES_REQUESTED** [review]

**Review by evandroAlfieri — APPROVED** [review]

File: features/GEM051_Presence-of-geometric-context.feature (hunk 1 above)
Comment (aothms) [reply features/GEM051_Presence-of-geometric-context.feature]: I was a bit surprised by this, but see now that this is handled in https://github.com/buildingSMART/ifc-gherkin-rules/blob/f7dcaac762beb0c613d46f29fe94ca8aa778f9dd/features/steps/givens/attributes.py#L120

Just a heads up that IFC 4.4 is going to be available in the foreseeable future. We might want to do this a bit more generically. For inspiration see https://github.com/IfcOpenShell/IfcOpenShell/blob/6cb063ebd2e4f79b1c679b30e31af82d679b5cf5/src/ifcopenshell-python/ifcopenshell/file.py#L409-L420 We can create our own function that conditionally uses `"."` or `"X"` based on the major version.

But this is also a reminder that of course any inconsistency we introduce on ourselves is going to be have to be maintained indefinitely.

The good news is that everything applicable to IFC4.3 is almost for sure also applicable to IFC4.4 so we can just globally replace `IFC4.3` with `IFC4.3 or IFC4.4`.

**Pull request merged**